Add Sorting Functionality for Quick Search Results

SougandhS opened the pull request:
This PR adds sorting functionality to the Quick Search results, enabling users to sort entries by Line Number, Text, or by Path. The update introduces a user interface element (such as a dropdown or clickable column headers) that allows users to choose their preferred sorting method. This improves the usability of the search feature, especially when working with large result sets, by helping users locate relevant entries more quickly.


https://github.com/user-attachments/assets/63b6dcf1-a30d-460e-8fdd-5caf652675d4



Fixes : https://github.com/eclipse-platform/eclipse.platform.ui/issues/1940

SougandhS committed:
Add Sorting Functionality for Quick Search Results

This commit adds sorting functionality to the Quick Search results,
allowing users to sort entries by Line Number, Text, and Path.

Fixes : https://github.com/eclipse-platform/eclipse.platform.ui/issues/1940

SougandhS commented:
Hi @mickaelistria 
Could you please check this PR when you have some time..

SougandhS commented:
Hi @merks, could you please take a look at this PR when you get a sec?

SougandhS commented:
hi @iloveeclipse, could you please review this ?

merks reviewed (commented):
I’m traveling and looking with an iPhone. It’s not clear why some of the fields are arrays.  Mostly I see access via [0].

SougandhS commented:
> I’m traveling and looking with an iPhone. It’s not clear why some of the fields are arrays. Mostly I see access via [0].

Thanks Ed Merks for checking. I used single-element final arrays here so I could retain and update the values inside `TableColumn.addSelectionListener `'s new `SelectionAdapter() {...}.().` class.

merks commented:
> > I’m traveling and looking with an iPhone. It’s not clear why some of the fields are arrays. Mostly I see access via [0].
> 
> Thanks Ed Merks for checking. I used single-element final arrays here so I could retain and update the values inside `TableColumn.addSelectionListener `'s new `SelectionAdapter() {...}.().` class.

So why not ordinary non-final fields? Taking off now.

SougandhS commented:
> So why not ordinary non-final fields? Taking off now.

non-final values can work too, I will update it

iloveeclipse reviewed (changes requested):
I see (smaller) style issues, but unfortunately too many of them to approve:

1. Arrays passed to access first element only to `handleColumnSort()`. Looks ugly, hard to debug. Cant we have `sortColumnIndex` and `sortDirectionAscending` just as regulat fields?
2. New code is not properly formatted (no spaces between arguments, in if statements etc). Looks really not nice. E.g. `if(comparator==null)` should be `if (comparator == null)`.
3. `contentProvider.sortList();` called in the middle of unrelated code in `refreshWidgets()`. Can't it be called after `list.setItemCount(itemCount);` ?
4. `items.sort(comparator);` is executed on each call of `getElements()`. Shouldn't be called **once** at the place where the sort order is initialized / changed?

SougandhS commented:
>I see (smaller) style issues, but unfortunately too many of them to approve:

1. Done, refactored accordingly
2. Done
3. Done, removed from `refreshWidgets()` -> didn't required when 4. was done
4. Done, moved to block where `comparator` is set

iloveeclipse commented on bundles/org.eclipse.text.quicksearch/src/org/eclipse/text/quicksearch/internal/ui/QuickSearchDialog.java:
```diff
@@ -1551,6 +1600,26 @@ public void updateElement(int index) {
+		 */
+		public void setComparator(Comparator<LineItem> comparator) {
+			this.comparator = comparator;
+			items.sort(comparator);
```
`items.sort(comparator);` is not needed or expected here according to method purpose/javadoc. 
The caller side does that already after setting comparator. Please remove this line.

SougandhS replied:
Okay, I have removed it 👍

iloveeclipse commented on bundles/org.eclipse.text.quicksearch/src/org/eclipse/text/quicksearch/internal/ui/QuickSearchDialog.java:
```diff
@@ -840,19 +844,45 @@ public void getName(AccessibleEvent e) {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				Comparator<LineItem> lineNumberComparator = Comparator.comparingInt(LineItem::getLineNumber);
+				handleColumnSort( table, 0, lineNumberComparator);
```
Please remove extra in space `handleColumnSort( table`

SougandhS replied:
done

iloveeclipse commented on bundles/org.eclipse.text.quicksearch/src/org/eclipse/text/quicksearch/internal/ui/QuickSearchDialog.java:
```diff
@@ -840,19 +844,45 @@ public void getName(AccessibleEvent e) {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				Comparator<LineItem> textComparator = Comparator.comparing( LineItem::getText);
+				handleColumnSort( table, 1, textComparator);
```
Please remove extra in space `handleColumnSort( table`

SougandhS replied:
done

iloveeclipse commented on bundles/org.eclipse.text.quicksearch/src/org/eclipse/text/quicksearch/internal/ui/QuickSearchDialog.java:
```diff
@@ -840,19 +844,45 @@ public void getName(AccessibleEvent e) {
+		column.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				Comparator<LineItem> lineItemComparator = Comparator.comparing( item -> item.getFile().getFullPath().toString());
```
Please remove extra in space `Comparator.comparing( item`

SougandhS replied:
Done

iloveeclipse commented on bundles/org.eclipse.text.quicksearch/src/org/eclipse/text/quicksearch/internal/ui/QuickSearchDialog.java:
```diff
@@ -840,19 +844,45 @@ public void getName(AccessibleEvent e) {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				Comparator<LineItem> lineItemComparator = Comparator.comparing( item -> item.getFile().getFullPath().toString());
+				handleColumnSort( table, 2,lineItemComparator);
```
Please remove extra in space `handleColumnSort( table`
and add space in `2,lineItemComparator);`

SougandhS replied:
done

SougandhS commented:
Hi @iloveeclipse
Could you please check this again ?

SougandhS commented:
Hi @merks 
Could you please check this ?

merks commented on bundles/org.eclipse.text.quicksearch/src/org/eclipse/text/quicksearch/internal/ui/QuickSearchDialog.java:
```diff
@@ -951,6 +981,25 @@ public void keyPressed(KeyEvent e) {
+			contentProvider.setComparator(sorter);
+		} else {
+			contentProvider.setComparator(sorter.reversed());
+		}
```
It's a nit, but this could be a singe call to `setComparator(sortDirectionAscending ? sorter : sorter.reversed())`

SougandhS replied:
Thanks, updated

merks reviewed (approved):
Look okay.  Just one small improvement could be applied.

SougandhS commented:
> Look okay. Just one small improvement could be applied.

If there are no more issues, can this be merged ?

merks commented:
I’m not sure if @iloveeclipse still wants to review.

iloveeclipse commented:
No, I'm out of PC till 26.08

SougandhS commented:
Thank you @iloveeclipse & @merks

SougandhS commented:
Hi @merks 
I saw this feature being mentioned in 4.37 Release video (https://www.youtube.com/watch?v=xpyqX-AxEP0). Should I include this in N&N of 4.37 ?

merks commented:
Sure. 👍
